File: js/2-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    const head = new ListNode(0);
    let current = head, node1 = l1, node2 = l2, carry = 0;
    while (node1 || node2 || carry) {
        const v = (node1?.val ?? 0) + (node2?.val ?? 0) + carry;
        carry = 0;
        if(v > 9) carry++;
        current.next = new ListNode(v % 10);
        node1 = node1?.next;
        node2 = node2?.next;
        current = current.next;
    };

    return head.next;
};

File: python/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        # current = head, n1 = l1, n2 = l2, carry=0;
        current, n1, n2, carry = head, l1, l2, 0
        while n1 or n2 or carry:
            v = (n1.val if n1 else 0) + (n2.val if n2 else 0) + carry
            carry = v // 10
            current.next = ListNode(v % 10)
            n1=(n1.next if n1 else None)
            n2=(n2.next if n2 else None)
            current=current.next

        return head.next
